### PR TITLE
Add Blockly option to the level types dropdown for filtering

### PR DESCRIPTION
**Pull request by dmcavoy**
Follow up to https://github.com/code-dot-org/code-dot-org/pull/42916. Added Blockly as an option to the level types dropdown to allow editors to filter for Blockly specific levels. 

https://user-images.githubusercontent.com/208083/137204705-e85e9d98-7b0d-4268-aca9-e896ca58c4a2.mov

## Links

Previous PR: https://github.com/code-dot-org/code-dot-org/pull/42916
Jira: https://codedotorg.atlassian.net/browse/PLAT-1375

## Testing story

- Manual testing shown in the video

### CODE REVIEW

File: dashboard/app/models/levels/level.rb [hunk 1]
```diff
@@ -742,7 +742,7 @@ def self.search_options
     {
       levelOptions: [
         ['All types', ''],
-        *LevelsController::LEVEL_CLASSES.map {|x| [x.name, x.name]}.sort_by {|a| a[0]}
+        *LevelsController::LEVEL_CLASSES.map {|x| [x.name, x.name]}.push(['Blockly', 'Blockly']).sort_by {|a| a[0]}
```
Comment (tim-dot-org) [inline dashboard/app/models/levels/level.rb:745]: is there a reason we can't just add to the LEVEL_CLASSES constant?

**Comment by davidsbailey** [comment]
> Follow up to #42916. Added Blockly as an option to the level types dropdown to allow editors to filter for Blockly specific levels.

I don't love the user experience shown in the video when searching for "3_1". why not let the curriculum writer search by the entire level key instead?

File: dashboard/app/models/levels/level.rb (hunk 1 above)
Comment (dmcavoy) [reply dashboard/app/models/levels/level.rb:745]: The comment above LEVEL_CLASSES says that it is the list of level types you can make new levels of. I don't think we want new blockly levels made.

**Comment by dmcavoy** [comment]
> > Follow up to #42916. Added Blockly as an option to the level types dropdown to allow editors to filter for Blockly specific levels.
> 
> I don't love the user experience shown in the video when searching for "3_1". why not let the curriculum writer search by the entire level key instead?

This seemed like the most consistent flow to what we do for other levels. I think one of the things I'm wondering about is once we migrate all scripts off of script DSL where you would go to find the level key. If you go to a level it just says "blockly" in the gray box. I was wondering if we should update that gray box for these levels to give the level_num and then I feel like the experience would match what you expect on other levels.

Screen Shot 2021-10-14 at 9 51 12 AM

**Comment by dmcavoy** [comment]
I guess digging deeper we show the level key on the new UI 

image

So I guess my question is are you adding the level by level key? Is that what we are searching when searching name?

**Comment by davidsbailey** [comment]
if the curriculum writer had to use the UX currently shown in this PR 3 months from now, I doubt they'd get through the process of adding a blockly level without asking for help. However if we always show them the full level key for blockly levels, then they should have no problem figuring out how to search for it when they need to add it.

> If you go to a level it just says "blockly" in the gray box.

Good catch! This should be easy to update to show the level key.

> So I guess my question is are you adding the level by level key? Is that what we are searching when searching name?

my perspective here is to focus on the user. tying the UI to the underlying implementation is not the goal.

In the UI, yes, I'm suggesting we present an interface where the user searches by level key.

**Comment by davidsbailey** [comment]
Happy to discuss at standup today. this could be a good welcome-back topic for @tess323 😁

**Comment by tess323** [comment]
@davidsbailey making sure I understand fully - is your recommendation to 

1. Add the level key to the gray box on the level, so writers have an easy way to know the key
2. add a search box specifically for level key to the UI?

My only concern with that is horizontal space, but I think since its an edit UI I am okay if the search flows into two lines. 

One question that I have for curriculum writers -- do writers search for other types of levels by level key or is this Blockly specific?

**Comment by davidsbailey** [comment]
> @davidsbailey making sure I understand fully - is your recommendation to
> 
> 1. Add the level key to the gray box on the level, so writers have an easy way to know the key
> 2. add a search box specifically for level key to the UI?
> 
> My only concern with that is horizontal space, but I think since its an edit UI I am okay if the search flows into two lines.

Sorry, no, I'm suggesting that they paste in the level key into the same search box, and we use the fact that it starts with "blockly:" to find it in a different way.

> One question that I have for curriculum writers -- do writers search for other types of levels by level key or is this Blockly specific?

level name == level key for non-blockly levels, so yes, this would mean that curriculum writers are now searching for all levels by level key. level key is the definitive token used to include a level into a script.

**Comment by dmcavoy** [comment]
After discussion in stand up we are going to use the level key as the name of the level everywhere for blockly levels. This means we will need to update the name that shows in the table for the AddLevelsDialog and the name that shows in the gray box on the level. I will also update the search to look for level key.

File: dashboard/app/models/levels/level.rb [hunk 2]
```diff
@@ -594,7 +594,7 @@ def summarize_for_edit
     {
       id: id.to_s,
       type: self.class.to_s,
-      name: name == 'blockly' ? level_num : name,
+      name: name == 'blockly' ? key : name,
```
Comment (davidsbailey) [inline dashboard/app/models/levels/level.rb]: I can't quite explain it, but this is unfortunately not a reliable way to determine whether a level is a blockly level:
```
[levelbuilder] dashboard > Level.where.not(level_num: ['custom', nil]).where.not(name: 'blockly').map(&:key)
=> ["blockly:Jigsaw:1", "blockly:Flappy:11", "blockly:Flappy:2", "blockly:MazeStep:2_1_step", "blockly:Maze:2_15", "blockly:Maze:2_16", "blockly:Maze:2_17", "blockly:MazeStep:2_17_step", "blockly:Maze:2_18", "blockly:Maze:2_19", "blockly:MazeStep:2_2_step", "blockly:Flappy:3", "blockly:MazeEC:3_1", "blockly:MazeEC:3_2", "blockly:MazeEC:3_3", "blockly:MazeEC:3_4", "blockly:Flappy:4", "blockly:Flappy:5", "blockly:Flappy:6", "blockly:Flappy:7", "blockly:Flappy:8", "blockly:Custom:builder", "blockly:ArtistEC:ec_1_10", "blockly:ArtistEC:ec_1_2", "blockly:StudioEC:ec_sandbox", "blockly:Webapp:ec_simple", "blockly:Eval:eval1", "blockly:Calc:example1", "blockly:Studio:full_sandbox", "blockly:Studio:full_sandbox_infinity", "blockly:MazeStep:karel_1_9_step", "blockly:MazeStep:karel_2_9_step", "blockly:MazeStep:karel_bee_1"]
```
you should just be able to use `name: key` here.

File: dashboard/app/views/levels/_admin.html.haml [hunk 3]
```diff
@@ -1,6 +1,9 @@
 - if current_user && (current_user.levelbuilder? || current_user.project_validator? || Experiment.get_editor_experiment(current_user))
   = render layout: 'shared/extra_links' do
-    %strong= @level.name
+    - if @level.name == 'blockly'
+      %strong= @level.key
+    - else
+      %strong= @level.name
```
Comment (davidsbailey) [inline dashboard/app/views/levels/_admin.html.haml]: I think you need just `@level.key` here

**Review by davidsbailey — APPROVED** [review]
Thank you for adopting this approach, Dani! Just one more annoying edge case to deal with, then I think this looks good to go.

File: dashboard/app/models/levels/level.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/models/levels/level.rb]: more specifically, the part I can explain is that there are some blockly levels (i.e. level num is not "custom" or nil) which are not named "blockly". the part I can't explain is why they are not named "blockly".

Comment (dmcavoy) [reply dashboard/app/models/levels/level.rb]: Does this impact the searching where you start the search with 'blockly'?

Comment (dmcavoy) [reply dashboard/app/models/levels/level.rb]: Looks like they all start with blockly but just checking

Comment (davidsbailey) [reply dashboard/app/models/levels/level.rb]: > Does this impact the searching where you start the search with 'blockly'?

all blockly levels have a level key starting with `blockly:`, so we should be good there.

**Pull request merged**